libraries/adaptive-expressions/src/functionUtils.ts

```diff
@@ -0,0 +1,1033 @@
+                        soFar = value;
+                    }
+
+                }
+
+                return {value: soFar, error: undefined};
+            },
+            verify
+        );
+    }
+
+    /**
+     * Lookup a property in IDictionary, JObject or through reflection.
+     * @param instance Instance with property.
+     * @param property Property to lookup.
+     * @returns Value and error information if any.
+     */
+    public static accessProperty(instance: any, property: string): {value: any; error: string} {
+        // NOTE: This returns null rather than an error if property is not present
+        if (!instance) {
+            return {value: undefined, error: undefined};
+        }
+
+        let value: any;
+        let error: string;
+        // todo, Is there a better way to access value, or any case is not listed below?
+        if (instance instanceof Map && instance as Map<string, any> !== undefined) {
+            const instanceMap: Map<string, any> = instance as Map<string, any>;
+            value = instanceMap.get(property);
+            if (value === undefined) {
+                const prop: string = Array.from(instanceMap.keys()).find((k: string): boolean => k.toLowerCase() === property.toLowerCase());
+                if (prop !== undefined) {
+                    value = instanceMap.get(prop);
+                }
+            }
+        } else {
+            const prop: string = Object.keys(instance).find((k: string): boolean => k.toLowerCase() === property.toLowerCase());
+            if (prop !== undefined) {
+                value = instance[prop];
+            }
+        }
+
+        return {value, error};
+    }
+
+    /**
+     * Lookup a property in IDictionary, JObject or through reflection.
+     * @param instance Instance with property.
+     * @param property Property to lookup.
+     * @returns Value and error information if any.
+     */
+    public static accessIndex(instance: any, index: number): {value: any; error: string} {
+        // NOTE: This returns null rather than an error if property is not present
+        if (instance === null || instance === undefined) {
+            return {value: undefined, error: undefined};
+        }
+
+        let value: any;
+        let error: string;
+
+        if (Array.isArray(instance)) {
+            if (index >= 0 && index < instance.length) {
+                value = instance[index];
+            } else {
+                error = `${index} is out of range for ${instance}`;
+            }
+        } else {
+            error = `${instance} is not a collection.`;
+        }
+
+        return {value, error};
+    }
+
+    /**
+     * transform a timestamp into another with customized function.
+     * @param timeStamp Original time stamp.
+     * @param transform Transform function.
+     * @returns new timestamp and error.
+     */
+    public static parseTimestamp(timeStamp: string, transform?: (arg0: Date) => any): {value: any; error: string} {
```

Comment:
>{value: any; error: string} [](start = 86, length = 27)

Turn this into a typescript type and use it everywhere.  Less error prone and clearer code.